### ISSUE TITLE
FSDP/veomni/torchtitan + actor.checkpoint.async_save=True never writes latest_checkpointed_iteration.txt, so resume_mode=auto silently restarts from scratch

### ISSUE THREAD

**Issue by YeonwooSung** [verl-project/verl#7952]
### System Info

- verl: `main` @ `23a341e3` ([model] chore: add Qwen3.8-27B Megatron GRPO script (#7717))
- The failing code path is CPU-only control flow; reproduced locally on macOS / Python 3.12 / torch 2.14 CPU. No GPU is needed to see it.

### Information

- [x] My own modified scripts

### Tasks

- [x] My own task or dataset (give details below)

### Reproduction

**Trigger config:** any non-Megatron training strategy (`actor_rollout_ref.actor.strategy` in `fsdp` / `fsdp2` / `veomni` / `torchtitan`) combined with
`actor_rollout_ref.actor.checkpoint.async_save=True` and `trainer.resume_mode=auto`.

**What happens:** the resume tracker `latest_checkpointed_iteration.txt` is never written by anyone, so on restart the trainer silently starts from step 0 and overwrites the run. There is no warning and no error.

#### The code path

Both trainers unconditionally skip the tracker write whenever `async_save` is truthy, on the assumption that the checkpoint manager will write it later:

`verl/trainer/ppo/v1/trainer_base.py:1025-1041`
```python
# write latest checkpointed iteration tracker for atomic resume
actor_ckpt_cfg = self.config.actor_rollout_ref.actor.get("checkpoint", {})
if actor_ckpt_cfg.get("async_save", False):
    logger.info("skip write latest_checkpointed_iteration.txt when async_save is True")
    self.checkpoint_callback.on_save(..., async_save=True)
    return
local_latest_checkpointed_iteration = os.path.join(
    self.config.trainer.default_local_dir, "latest_checkpointed_iteration.txt"
)
```

`verl/trainer/ppo/ray_trainer.py:1036-1050` has the same branch (v0).

Neither branch looks at `actor.strategy`. That assumption only holds for Megatron:

```
$ grep -c async_save verl/utils/checkpoint/megatron_checkpoint_manager.py
10
$ grep -c async_save verl/utils/checkpoint/fsdp_checkpoint_manager.py
0
```

`MegatronCheckpointManager` does write the tracker itself (`verl/utils/checkpoint/megatron_checkpoint_manager.py:1136-1146`, guarded by `self.checkpoint_config.async_save and self.rank == 0`).
`FSDPCheckpointManager` — which backs `fsdp`, `fsdp2` (`verl/workers/engine/fsdp/transformer_impl.py:217`) and `veomni` (`verl/workers/engine/veomni/transformer_impl.py:253`) — has no `async_save` handling at all. It saves synchronously and never touches the tracker. So for every non-Megatron strategy the tracker write is skipped by the trainer and picked up by nobody.

`verl/trainer/config/config.py:37` already documents the intended scope:

```python
async_save (bool): Whether to save checkpoints asynchronously. Only implemented for Megatron as of now.
```

...but nothing enforces it, and `async_save: False` is present in `verl/trainer/config/actor/actor.yaml:142` for every backend, so it is a reachable knob on FSDP.

#### Executed reproduction

Script (drives the real `RayPPOTrainer._save_checkpoint`; I used the v0 trainer because importing `verl.trainer.ppo.v1` needs `transfer_queue`, which is not installed in my CPU env — the v1 branch is byte-for-byte equivalent):

```python
import os, tempfile
from unittest.mock import MagicMock
from omegaconf import OmegaConf
from verl.trainer.ppo.ray_trainer import RayPPOTrainer
from verl.utils.checkpoint.checkpoint_manager import find_latest_ckpt_path

def run(async_save):
    tmp = tempfile.mkdtemp()
    t = RayPPOTrainer.__new__(RayPPOTrainer)
    t.global_steps, t.use_critic = 3, False
    t.config = OmegaConf.create({
        "trainer": {"default_local_dir": tmp, "default_hdfs_dir": None},
        "actor_rollout_ref": {"actor": {"strategy": "fsdp",
                                        "checkpoint": {"async_save": async_save}}},
    })
    t.actor_rollout_wg = MagicMock()
    # emulate FSDPCheckpointManager: writes shards synchronously, nothing else
    def _save(local_path, *a, **k):
        os.makedirs(local_path, exist_ok=True)
        open(os.path.join(local_path, "model_world_size_1_rank_0.pt"), "w").close()
    t.actor_rollout_wg.save_checkpoint.side_effect = _save
    dl = MagicMock(); dl.state_dict.return_value = {}
    t.train_dataloader = dl
    t._save_checkpoint()
    print(f"  async_save={async_save}")
    print(f"    dir contents          : {sorted(os.listdir(tmp))}")
    print(f"    find_latest_ckpt_path : {find_latest_ckpt_path(tmp)}")

run(False)
run(True)
```

Output:

```
strategy=fsdp  (FSDPCheckpointManager -- 0 occurrences of 'async_save')

  async_save=False
    dir contents          : ['global_step_3', 'latest_checkpointed_iteration.txt']
    find_latest_ckpt_path : /var/folders/.../tmpr6r26keo/global_step_3

skip write latest_checkpointed_iteration.txt when async_save is True
  async_save=True
    dir contents          : ['global_step_3']
    Checkpoint tracker file does not exist: /var/folders/.../tmpii4o5hdy/latest_checkpointed_iteration.txt
    find_latest_ckpt_path : None
```

The checkpoint data is fully on disk under `global_step_3/`; only the pointer to it is missing.

#### Consequence on resume

`find_latest_ckpt_path` (`verl/utils/checkpoint/checkpoint_manager.py:222-238`) returns `None` as soon as the tracker file is absent. Its callers then treat that as "no checkpoint":

- `verl/trainer/ppo/v1/trainer_base.py:875-878` → `logger.info("Training from scratch"); return`
- `verl/trainer/ppo/ray_trainer.py:1064-1069` → `print("Training from scratch"); return 0`

So a long FSDP run that saved every N steps restarts at step 0 on the next launch. The checkpoints are intact on disk but unreachable via `resume_mode=auto`; the run then overwrites them from scratch. Nothing in the logs flags it beyond a single INFO line.

**What I verified by execution:** the two `grep -c` counts, the repro above, and `find_latest_ckpt_path` returning `None`.
**What I verified by reading:** the v1 trainer branch (module not importable in a CPU-only env), the `_load_checkpoint` "Training from scratch" paths, the Megatron manager's tracker write, and which engines construct `FSDPCheckpointManager`.

**Not affected:** SFT. `verl/trainer/sft_trainer.py` goes through `CheckpointHandler`, which writes the tracker unconditionally (`verl/utils/checkpoint/checkpoint_handler.py:128`) and has no `async_save` branch.

### Expected behavior

Either the combination is rejected up front, or the tracker gets written. Concretely, two directions — I'd like to know which one you prefer before I push anything beyond the guard:

**(a) Fail closed at config validation.** Raise in `ActorConfig.__post_init__` when `checkpoint.async_save` is set and `strategy != "megatron"`, with a message pointing at the Megatron-only note in `config.py:37`. Smallest change, matches the documented scope, covers v0 / v1 / any future trainer in one place, and turns a silent whole-run data loss into a startup error. Downside: it is a hard error for anyone who has the flag set today (they are already silently broken).

**(b) Make the non-Megatron path write the tracker.** Gate the skip on `strategy == "megatron"` in both trainers instead of on `async_save` alone. Nothing breaks for existing users, and it is correct today because `FSDPCheckpointManager` saves synchronously — but it silently normalises a flag that does nothing, and it would have to be revisited the moment FSDP grows a real async writer.

Related, and the reason I'm asking rather than just sending (b): **#4311** (`[fsdp, doc] feat: Enhance FSDP checkpointing with DCP support`, open since 2025-11-26, last touched 2026-01-05) implements DCP-based async save for FSDP and flips the `config.py:37` docstring to "supported for both FSDP and Megatron". If that lands, (a)'s guard is deleted in one line and (b) becomes wrong again. If it is not landing soon, `main` keeps silently discarding runs. Happy to send whichever you'd rather carry.

I have a fix for (a) ready with a CPU unit test and can open the PR immediately.

**Comment by YeonwooSung** [comment]
Opened #7955 with direction (a) (reject the combination in `ActorConfig.__post_init__`), plus the CPU unit tests. If you'd rather have direction (b), or would rather wait on #4311, say so on the PR and I'll rework or close it.
